R/susie_vbupdate_ss.R

```diff
@@ -3,33 +3,33 @@
 #' @param Xty a p vector
 #' @param s_init a list with elements sigma2, V, alpha, mu, Xr
 #' @param estimate_prior_variance boolean indicating whether to estimate prior variance
+#' @importFrom Matrix diag
 update_each_effect_ss <- function (XtX, Xty, s_init, estimate_prior_variance=FALSE) {
 
   # Repeat for each effect to update
   s = s_init
   L = nrow(s$alpha)
+  dXtX = Matrix::diag(XtX)
 
   if(L>0){
     for (l in 1:L){
       # remove lth effect from fitted values
-      s$XtXr = s$XtXr - XtX %*% (s$alpha[l,] * s$mu[l,])
+      s$XtXr = s$XtXr - as(XtX %*% (s$alpha[l,] * s$mu[l,]), 'matrix')
 
       #compute residuals
       XtR = Xty - s$XtXr
-
-      res = single_effect_regression_ss(XtR,diag(XtX),s$V[l],s$sigma2,estimate_prior_variance)
-
+      res = single_effect_regression_ss(XtR,dXtX,s$V[l],s$sigma2,estimate_prior_variance)
       # Update the variational estimate of the posterior mean.
       s$mu[l,] <- res$mu
       s$alpha[l,] <- res$alpha
       s$mu2[l,] <- res$mu2
       s$V[l] <- res$V
-      s$KL[l] <- -res$logBF + SER_posterior_e_loglik_ss(XtX,XtR,s$sigma2,res$alpha*res$mu,res$alpha*res$mu2)
+      s$KL[l] <- -res$logBF + SER_posterior_e_loglik_ss(dXtX,XtR,s$sigma2,res$alpha*res$mu,res$alpha*res$mu2)
 
-      s$XtXr <- s$XtXr + XtX %*% (s$alpha[l,]*s$mu[l,])
+      s$XtXr <- s$XtXr + as(XtX %*% (s$alpha[l,]*s$mu[l,]), 'matrix')
     }
   }
-
+  s$XtXr = unname(s$XtXr)
```

Comment:
Do you think that instead of doing: 

```
for i in L:
   xtxr += as.matrix(x)
xtxr = unname(xtxr)
```

it is cleaner to do:

```
for i in L:
   xtxr += x
xtxr = unname(as.matrix(xtxr))
```

?